File: webapp/backend/greenit.py
"""
greenit.py — lecture et agrégation « efficience » du grand livre `api_usage.log`.

Pourquoi un module dédié plutôt qu'une extension de services.py :

1. **Parsing défensif volontaire.** Le chantier GreenIT enrichit le ledger avec
   des champs OPTIONNELS (`octets_entrants`, `octets_sortants`, `duree_ms`,
   `energie_wh`, `co2e_g`, `modele`, `profil`). Les anciennes lignes ne les ont
   pas, et `diagnostic.api_schema.LedgerEntry` est en `extra="forbid"` : valider
   avec lui ferait *rejeter* les lignes enrichies tant que le schéma amont n'a
   pas atterri. On lit donc le JSONL brut et on applique des défauts (0 / None).
   Conséquence : cette vue fonctionne avec un ledger absent, ancien, enrichi, ou
   panaché des trois.

2. **Suivi incrémental.** Le flux SSE a besoin d'un tail par offset d'octets,
   qui n'a rien à faire dans une couche d'agrégats.

STRICTEMENT EN LECTURE : ce module ouvre `api_usage.log` en lecture seule. Aucune
écriture, aucun réseau, aucun `api_io`, aucune transition d'état.

⚠️ Les valeurs `energie_wh` et `co2e_g` sont des ESTIMATIONS issues des facteurs
paramétrables de `knowledge/greenit.yaml` (ordres de grandeur, pas des mesures
certifiées). Toute réponse de ce module les étiquette comme telles.
"""
from __future__ import annotations

import json
import os
from datetime import date, datetime, timezone
from pathlib import Path
from typing import Any, Iterable

# Champs GreenIT ajoutés au ledger par le chantier d'instrumentation. Ils sont
# tous facultatifs : `normaliser()` garantit une forme complète en sortie.
CHAMPS_NUMERIQUES_GREENIT = (
    "octets_entrants",
    "octets_sortants",
    "duree_ms",
    "energie_wh",
    "co2e_g",
)

NOTE_ESTIMATION = (
    "energie_wh et co2e_g sont des ESTIMATIONS calculées à partir des facteurs "
    "paramétrables de knowledge/greenit.yaml (ordres de grandeur destinés à "
    "comparer des scénarios). Ce ne sont pas des mesures certifiées et elles ne "
    "constituent pas un bilan carbone opposable."
)


# ---------------------------------------------------------------------------
# Normalisation d'une ligne
# ---------------------------------------------------------------------------

def _nombre(valeur: Any, defaut: float = 0.0) -> float:
    """Convertit en float ; tout ce qui n'est pas numérique retombe sur `defaut`.

    Le ledger est un fichier append-only écrit par plusieurs agents : une valeur
    `null`, une chaîne ou un champ manquant ne doit jamais faire tomber la vue.
    """
    if valeur is None or isinstance(valeur, bool):
        return defaut
    if isinstance(valeur, (int, float)):
        return float(valeur)
    try:
        return float(str(valeur))
    except (TypeError, ValueError):
        return defaut


def _texte(valeur: Any) -> str | None:
    if valeur is None:
        return None
    texte = str(valeur).strip()
    return texte or None


def normaliser(obj: dict) -> dict:
    """Ramène une ligne JSONL brute à une forme complète et typée.

    Les champs GreenIT absents valent 0 (numériques) ou None (modele/profil) :
    une ancienne ligne produit donc une empreinte nulle plutôt qu'une erreur.
    """
    unites_brutes = obj.get("unites")
    unites: dict[str, float] = {}
    if isinstance(unites_brutes, dict):
        for cle, val in unites_brutes.items():
            unites[str(cle)] = _nombre(val)

    # « tokens » = toute unité dont le nom contient "token" (input/output/cache).
    tokens = sum(v for k, v in unites.items() if "token" in k.lower())

    ligne = {
        "ts": _texte(obj.get("ts")),
        "fournisseur": _texte(obj.get("fournisseur")) or "inconnu",
        "endpoint": _texte(obj.get("endpoint")) or "",
        "unites": unites,
        "tokens": tokens,
        "cout_estime": _nombre(obj.get("cout_estime")),
        "devise": _texte(obj.get("devise")) or "USD",
        "fiche": _texte(obj.get("fiche")),
        "cache_hit": bool(obj.get("cache_hit", False)),
        "resultat": _texte(obj.get("resultat")) or "ok",
        "detail": obj.get("detail") or "",
        # --- champs GreenIT (optionnels côté ledger) ---
        "modele": _texte(obj.get("modele")),
        "profil": _texte(obj.get("profil")),
    }
    for champ in CHAMPS_NUMERIQUES_GREENIT:
        ligne[champ] = _nombre(obj.get(champ))
    ligne["octets"] = ligne["octets_entrants"] + ligne["octets_sortants"]
    # Vrai seulement si la ligne porte au moins un champ d'empreinte : permet
    # d'afficher honnêtement le taux de couverture de l'instrumentation.
    ligne["greenit_instrumente"] = any(
        champ in obj and obj.get(champ) is not None
        for champ in CHAMPS_NUMERIQUES_GREENIT
    )
    return ligne


def _date_de_ligne(ligne: dict) -> str:
    """Extrait la partie AAAA-MM-JJ du ts (comparable lexicographiquement)."""
    ts = ligne.get("ts") or ""
    return ts[:10]


# ---------------------------------------------------------------------------
# Lecture du ledger
# ---------------------------------------------------------------------------

def lire_ledger(path: Path, depuis: date | None = None) -> tuple[list[dict], int]:
    """Lit `api_usage.log` (JSONL). Retourne (lignes normalisées, nb illisibles).

    Ne lève jamais : fichier absent → ([], 0) ; ligne corrompue → ignorée et
    comptée. C'est un journal append-only lu à chaud, une ligne tronquée en
    cours d'écriture est un cas normal, pas une erreur.
    """
    if not path.is_file():
        return [], 0
    try:
        brut = path.read_text(encoding="utf-8", errors="replace")
    except OSError:
        return [], 0

    lignes: list[dict] = []
    illisibles = 0
    borne = depuis.isoformat() if depuis else None

    for raw in brut.splitlines():
        raw = raw.strip()
        if not raw:
            continue
        try:
            obj = json.loads(raw)
        except json.JSONDecodeError:
            illisibles += 1
            continue
        if not isinstance(obj, dict):
            illisibles += 1
            continue
        ligne = normaliser(obj)
        if borne is not None and _date_de_ligne(ligne) < borne:
            continue
        lignes.append(ligne)
    return lignes, illisibles


# ---------------------------------------------------------------------------
# Agrégation
# ---------------------------------------------------------------------------

_CHAMPS_CUMULES = ("cout", "energie_wh", "co2e_g", "octets", "duree_ms", "tokens")


def _bucket() -> dict[str, float]:
    b: dict[str, float] = {c: 0.0 for c in _CHAMPS_CUMULES}
    b["nb_appels"] = 0.0
    return b


def _cumuler(bucket: dict[str, float], ligne: dict) -> None:
    bucket["nb_appels"] += 1
    bucket["cout"] += ligne["cout_estime"]
    bucket["energie_wh"] += ligne["energie_wh"]
    bucket["co2e_g"] += ligne["co2e_g"]
    bucket["octets"] += ligne["octets"]
    bucket["duree_ms"] += ligne["duree_ms"]
    bucket["tokens"] += ligne["tokens"]


def _moyennes_par_cle(reels: Iterable[dict]) -> dict[tuple[str, str], dict[str, float]]:
    """Moyenne coût/énergie/CO2e des appels RÉELS, par (fournisseur, endpoint).

    Sert à estimer ce qu'un cache-hit a évité : on ne connaît pas le coût d'un
    appel qui n'a jamais eu lieu, on l'approche par la moyenne de ses jumeaux.
    """
    agg: dict[tuple[str, str], dict[str, float]] = {}
    for ligne in reels:
        for cle in (
            (ligne["fournisseur"], ligne["endpoint"]),
            (ligne["fournisseur"], "*"),
            ("*", "*"),
        ):
            b = agg.setdefault(cle, {"n": 0.0, "cout": 0.0, "energie_wh": 0.0, "co2e_g": 0.0})
            b["n"] += 1
            b["cout"] += ligne["cout_estime"]
            b["energie_wh"] += ligne["energie_wh"]
            b["co2e_g"] += ligne["co2e_g"]
    return agg


def agreger(lignes: list[dict], *, nb_illisibles: int = 0) -> dict:
    """Calcule l'agrégat d'efficience exposé par `GET /api/greenit`.

    Conventions alignées sur `diagnostic.usage.agreger` (contrat cohérent entre
    /api/usage et /api/greenit) :
      - un cache-hit n'est PAS un appel : il alimente `nb_cache_hits` ;
      - un appel `budget_depasse` compte dans `nb_appels` mais n'a rien dépensé,
        donc il n'alimente ni coût ni empreinte.
    """
    nb_appels = 0
    nb_cache_hits = 0
    nb_erreurs = 0
    nb_instrumentes = 0

    totaux = _bucket()
    par_modele: dict[str, dict[str, float]] = {}
    par_fournisseur: dict[str, dict[str, float]] = {}
    reels: list[dict] = []
    cache_hits: list[dict] = []

    for ligne in lignes:
        if ligne["cache_hit"]:
            nb_cache_hits += 1
            cache_hits.append(ligne)
            continue

        nb_appels += 1
        if ligne["resultat"] == "erreur":
            nb_erreurs += 1
        if ligne["resultat"] == "budget_depasse":
            # Refusé par le garde-fou : aucun octet n'a circulé.
            continue
        if ligne["greenit_instrumente"]:
            nb_instrumentes += 1

        reels.append(ligne)
        _cumuler(totaux, ligne)
        _cumuler(par_fournisseur.setdefault(ligne["fournisseur"], _bucket()), ligne)
        modele = ligne["modele"] or "non-llm"
        _cumuler(par_modele.setdefault(modele, _bucket()), ligne)

    # --- Économies attribuées au cache (ESTIMATION par moyenne des jumeaux) ---
    moyennes = _moyennes_par_cle(reels)
    eco_cout = eco_energie = eco_co2 = 0.0
    for hit in cache_hits:
        for cle in (
            (hit["fournisseur"], hit["endpoint"]),
            (hit["fournisseur"], "*"),
            ("*", "*"),
        ):
            b = moyennes.get(cle)
            if b and b["n"] > 0:
                eco_cout += b["cout"] / b["n"]
                eco_energie += b["energie_wh"] / b["n"]
                eco_co2 += b["co2e_g"] / b["n"]
                break
        # Aucun appel réel comparable → on ne peut rien estimer : 0, pas d'invention.

    total_tentatives = nb_appels + nb_cache_hits
    taux_cache = nb_cache_hits / total_tentatives if total_tentatives else 0.0

    def _liste(source: dict[str, dict[str, float]], cle_nom: str) -> list[dict]:
        out = []
        for nom, b in source.items():
            out.append(
                {
                    cle_nom: nom,
                    "nb_appels": int(b["nb_appels"]),
                    "tokens": round(b["tokens"], 3),
                    "cout": round(b["cout"], 6),
                    "energie_wh": round(b["energie_wh"], 6),
                    "co2e_g": round(b["co2e_g"], 6),
                    "octets": int(b["octets"]),
                    "duree_ms": round(b["duree_ms"], 3),
                }
            )
        # Tri par coût décroissant puis nom : ordre stable et lisible.
        out.sort(key=lambda d: (-d["cout"], d[cle_nom]))
        return out

    return {
        "cout_total_usd": round(totaux["cout"], 6),
        "devise": "USD",
        "energie_wh_total": round(totaux["energie_wh"], 6),
        "co2e_g_total": round(totaux["co2e_g"], 6),
        "octets_total": int(totaux["octets"]),
        "duree_ms_totale": round(totaux["duree_ms"], 3),
        "nb_appels": nb_appels,
        "nb_cache_hits": nb_cache_hits,
        "nb_erreurs": nb_erreurs,
        "nb_lignes_illisibles": nb_illisibles,
        "taux_cache": round(taux_cache, 4),
        "economies": {
            "appels_evites": nb_cache_hits,
            "cout_evite_usd": round(eco_cout, 6),
            "energie_wh_evitee": round(eco_energie, 6),
            "co2e_evite_g": round(eco_co2, 6),
            "methode": (
                "ESTIMATION — chaque cache-hit est valorisé à la moyenne des appels "
                "réels de même (fournisseur, endpoint). Sans appel réel comparable, "
                "l'économie est comptée à 0 plutôt qu'extrapolée."
            ),
        },
        "par_modele": _liste(par_modele, "modele"),
        "par_fournisseur": _liste(par_fournisseur, "fournisseur"),
        "estimation": True,
        "note_estimation": NOTE_ESTIMATION,
        "source_facteurs": "knowledge/greenit.yaml",
        "couverture_greenit": (
            round(nb_instrumentes / len(reels), 4) if reels else 0.0
        ),
        "ledger_present": True,  # écrasé par l'appelant si le fichier manque
    }


# ---------------------------------------------------------------------------
# Tail incrémental (support du flux SSE)
# ---------------------------------------------------------------------------

class LedgerTail:
    """Suit `api_usage.log` par offset d'octets, à la manière de `tail -f`.

    Points d'attention traités :
      - **ligne partielle** : une ligne sans `\\n` final est en cours d'écriture,
        on la garde en tampon et on ne la publie qu'une fois terminée ;
      - **troncature / rotation** : si le fichier rétrécit, on repart de 0 ;
      - **fichier absent** : offset 0, aucune ligne, aucune exception.
    """

    def __init__(self, path: Path, *, depuis_debut: bool = False) -> None:
        self.path = path
        self._tampon = ""
        self.offset = 0 if depuis_debut else self.taille()

    def taille(self) -> int:
        try:
            return os.path.getsize(self.path)
        except OSError:
            return 0

    def lire_nouvelles(self) -> list[dict]:
        """Retourne les lignes complètes apparues depuis le dernier appel."""
        taille = self.taille()
        if taille < self.offset:
            # Fichier tronqué ou remplacé : on recommence proprement.
            self.offset = 0
            self._tampon = ""
        if taille == self.offset:
            return []
        try:
            with self.path.open("r", encoding="utf-8", errors="replace") as fh:
                fh.seek(self.offset)
                bloc = fh.read()
                self.offset = taille
        except OSError:
            return []

        self._tampon += bloc
        *completes, self._tampon = self._tampon.split("\n")

        lignes: list[dict] = []
        for raw in completes:
            raw = raw.strip()
            if not raw:
                continue
            try:
                obj = json.loads(raw)
            except json.JSONDecodeError:
                continue
            if isinstance(obj, dict):
                lignes.append(normaliser(obj))
        return lignes


def dernieres_lignes(path: Path, n: int) -> list[dict]:
    """Les `n` dernières lignes normalisées (les plus anciennes d'abord)."""
    if n <= 0:
        return []
    lignes, _ = lire_ledger(path)
    return lignes[-n:]


def horodatage() -> str:
    return datetime.now(timezone.utc).isoformat()

File: webapp/frontend/src/lib/greenit.ts
// greenit.ts — fonctions PURES de l'écran d'efficience : parsing défensif des
// événements SSE, agrégation incrémentale côté client, formatage des unités.
//
// Pourquoi une agrégation côté client : le flux temps réel doit faire bouger
// les tuiles sans attendre un nouvel appel à /api/greenit. On applique donc
// chaque événement reçu sur le dernier agrégat connu, avec exactement les mêmes
// conventions que le backend (un cache-hit n'est pas un appel ; un appel
// `budget_depasse` n'a rien dépensé).
//
// ⚠️ energie_wh / co2e_g restent des ESTIMATIONS (facteurs de
// knowledge/greenit.yaml) — ces fonctions ne font que les additionner.

import type { Greenit, GreenitAppel } from "../types";

function nombre(v: unknown, defaut = 0): number {
  if (typeof v === "number" && Number.isFinite(v)) return v;
  if (typeof v === "string") {
    const n = Number(v);
    if (Number.isFinite(n)) return n;
  }
  return defaut;
}

function texte(v: unknown): string | null {
  if (typeof v !== "string") return null;
  const t = v.trim();
  return t === "" ? null : t;
}

/**
 * Normalise un événement `appel` du flux SSE.
 * Défensif par contrat : les champs GreenIT sont optionnels côté ledger, et
 * une charge utile inattendue ne doit jamais casser l'écran (→ null).
 */
export function parseAppel(brut: string | unknown): GreenitAppel | null {
  let obj: unknown = brut;
  if (typeof brut === "string") {
    try {
      obj = JSON.parse(brut);
    } catch {
      return null;
    }
  }
  if (obj === null || typeof obj !== "object" || Array.isArray(obj)) return null;
  const o = obj as Record<string, unknown>;

  const unites: Record<string, number> = {};
  if (o.unites && typeof o.unites === "object" && !Array.isArray(o.unites)) {
    for (const [k, v] of Object.entries(o.unites as Record<string, unknown>)) {
      unites[k] = nombre(v);
    }
  }
  const tokensCalcules = Object.entries(unites)
    .filter(([k]) => k.toLowerCase().includes("token"))
    .reduce((s, [, v]) => s + v, 0);

  const octetsEntrants = nombre(o.octets_entrants);
  const octetsSortants = nombre(o.octets_sortants);

  return {
    ts: texte(o.ts),
    fournisseur: texte(o.fournisseur) ?? "inconnu",
    endpoint: texte(o.endpoint) ?? "",
    unites,
    tokens: nombre(o.tokens, tokensCalcules),
    cout_estime: nombre(o.cout_estime),
    devise: texte(o.devise) ?? "USD",
    fiche: texte(o.fiche),
    cache_hit: o.cache_hit === true,
    resultat: texte(o.resultat) ?? "ok",
    detail: typeof o.detail === "string" ? o.detail : "",
    modele: texte(o.modele),
    profil: texte(o.profil),
    octets_entrants: octetsEntrants,
    octets_sortants: octetsSortants,
    octets: nombre(o.octets, octetsEntrants + octetsSortants),
    duree_ms: nombre(o.duree_ms),
    energie_wh: nombre(o.energie_wh),
    co2e_g: nombre(o.co2e_g),
    greenit_instrumente: o.greenit_instrumente === true,
  };
}

/** Clé d'identité d'un appel dans le flux (évite les doublons de rejeu). */
export function cleAppel(a: GreenitAppel, index: number): string {
  return `${a.ts ?? "?"}|${a.fournisseur}|${a.endpoint}|${index}`;
}

/**
 * Applique un appel reçu en direct sur l'agrégat courant (immutable).
 * Mêmes règles que le backend : cache-hit → pas un appel ; `budget_depasse` →
 * compté en appel mais sans coût ni empreinte (aucun octet n'a circulé).
 */
export function appliquerAppel(base: Greenit, appel: GreenitAppel): Greenit {
  const suivant: Greenit = {
    ...base,
    economies: { ...base.economies },
    par_modele: base.par_modele.map((m) => ({ ...m })),
    par_fournisseur: base.par_fournisseur.map((f) => ({ ...f })),
  };

  if (appel.cache_hit) {
    suivant.nb_cache_hits += 1;
    suivant.economies.appels_evites += 1;
    // Économie estimée = moyenne observée des appels réels déjà agrégés.
    if (suivant.nb_appels > 0) {
      suivant.economies.cout_evite_usd += suivant.cout_total_usd / suivant.nb_appels;
      suivant.economies.energie_wh_evitee +=
        suivant.energie_wh_total / suivant.nb_appels;
      suivant.economies.co2e_evite_g += suivant.co2e_g_total / suivant.nb_appels;
    }
    suivant.taux_cache = tauxCache(suivant.nb_appels, suivant.nb_cache_hits);
    return suivant;
  }

  suivant.nb_appels += 1;
  if (appel.resultat === "erreur") suivant.nb_erreurs += 1;
  suivant.taux_cache = tauxCache(suivant.nb_appels, suivant.nb_cache_hits);
  if (appel.resultat === "budget_depasse") return suivant;

  suivant.cout_total_usd += appel.cout_estime;
  suivant.energie_wh_total += appel.energie_wh;
  suivant.co2e_g_total += appel.co2e_g;
  suivant.octets_total += appel.octets;
  suivant.duree_ms_totale += appel.duree_ms;

  const modele = appel.modele ?? "non-llm";
  const ligneM = suivant.par_modele.find((m) => m.modele === modele);
  if (ligneM) {
    ligneM.nb_appels += 1;
    ligneM.tokens += appel.tokens;
    ligneM.cout += appel.cout_estime;
    ligneM.energie_wh += appel.energie_wh;
    ligneM.co2e_g += appel.co2e_g;
    ligneM.octets += appel.octets;
    ligneM.duree_ms += appel.duree_ms;
  } else {
    suivant.par_modele.push({
      modele,
      nb_appels: 1,
      tokens: appel.tokens,
      cout: appel.cout_estime,
      energie_wh: appel.energie_wh,
      co2e_g: appel.co2e_g,
      octets: appel.octets,
      duree_ms: appel.duree_ms,
    });
  }

  const ligneF = suivant.par_fournisseur.find(
    (f) => f.fournisseur === appel.fournisseur,
  );
  if (ligneF) {
    ligneF.nb_appels += 1;
    ligneF.tokens += appel.tokens;
    ligneF.cout += appel.cout_estime;
    ligneF.energie_wh += appel.energie_wh;
    ligneF.co2e_g += appel.co2e_g;
    ligneF.octets += appel.octets;
    ligneF.duree_ms += appel.duree_ms;
  } else {
    suivant.par_fournisseur.push({
      fournisseur: appel.fournisseur,
      nb_appels: 1,
      tokens: appel.tokens,
      cout: appel.cout_estime,
      energie_wh: appel.energie_wh,
      co2e_g: appel.co2e_g,
      octets: appel.octets,
      duree_ms: appel.duree_ms,
    });
  }

  suivant.par_modele.sort((a, b) => b.cout - a.cout || a.modele.localeCompare(b.modele));
  suivant.par_fournisseur.sort(
    (a, b) => b.cout - a.cout || a.fournisseur.localeCompare(b.fournisseur),
  );
  return suivant;
}

export function tauxCache(nbAppels: number, nbCacheHits: number): number {
  const total = nbAppels + nbCacheHits;
  return total > 0 ? nbCacheHits / total : 0;
}

// ---------------------------------------------------------------------------
// Formatage — unités physiques lisibles, sans jamais mentir sur la précision
// ---------------------------------------------------------------------------

/** Octets → Ko/Mo/Go (base 1024), 1 décimale au-delà du kilo. */
export function fmtOctets(n: number): string {
  const abs = Math.abs(n);
  if (abs < 1024) return `${Math.round(n)} o`;
  if (abs < 1024 ** 2) return `${(n / 1024).toFixed(1)} Ko`;
  if (abs < 1024 ** 3) return `${(n / 1024 ** 2).toFixed(1)} Mo`;
  return `${(n / 1024 ** 3).toFixed(2)} Go`;
}

/** Wh → mWh sous 1 Wh, kWh au-delà de 1000. */
export function fmtWh(n: number): string {
  const abs = Math.abs(n);
  if (abs === 0) return "0 Wh";
  if (abs < 1) return `${(n * 1000).toFixed(0)} mWh`;
  if (abs < 1000) return `${n.toFixed(2)} Wh`;
  return `${(n / 1000).toFixed(2)} kWh`;
}

/** Grammes de CO2e → g sous 1 kg, kg au-delà. */
export function fmtCo2(n: number): string {
  const abs = Math.abs(n);
  if (abs === 0) return "0 g";
  if (abs < 1) return `${(n * 1000).toFixed(0)} mg`;
  if (abs < 1000) return `${n.toFixed(2)} g`;
  return `${(n / 1000).toFixed(2)} kg`;
}

/** Millisecondes → ms / s / min. */
export function fmtDuree(ms: number): string {
  if (ms < 1000) return `${Math.round(ms)} ms`;
  if (ms < 60_000) return `${(ms / 1000).toFixed(1)} s`;
  return `${(ms / 60_000).toFixed(1)} min`;
}

/** Heure locale HH:MM:SS depuis un ts ISO (— si absent/illisible). */
export function fmtHeure(ts: string | null): string {
  if (!ts) return "—";
  const d = new Date(ts);
  if (Number.isNaN(d.getTime())) return ts.slice(11, 19) || "—";
  return d.toLocaleTimeString("fr-CA", { hour12: false });
}
